packages/shared/src/components/notifications/Toast.tsx

```diff
@@ -0,0 +1,164 @@
+import {
+  ToastNotification,
+  TOAST_NOTIF_KEY,
+} from '../../hooks/useToastNotification';
+import classed from '../../lib/classed';
+import { Button } from '../buttons/Button';
+import styles from './Toast.module.css';
+import XIcon from '../../../icons/x.svg';
+import { isTouchDevice } from '../../lib/tooltip';
+import {
+  NotifContainer,
+  NotifContent,
+  NotifMessage,
+  NotifProgress,
+} from './utils';
+
+interface ToastProps {
+  autoDismissNotifications?: boolean;
+}
+
+const Container = classed(NotifContainer, styles.toastContainer);
+const Progress = classed(NotifProgress, styles.toastProgress);
+const INTERVAL_COUNT = 10;
+const IN_OUT_ANIMATION = 140;
+const TEMPORARY_ID = 1;
+
+const Toast = ({
+  autoDismissNotifications = false,
+}: ToastProps): ReactElement => {
+  const router = useRouter();
+  const client = useQueryClient();
+  const [intervalId, setIntervalId] = useState<number>(null);
+  const [timer, setTimer] = useState(0);
+  const timeout = useRef<number>();
+  const { data: toast } = useQuery<ToastNotification>(TOAST_NOTIF_KEY);
+  const setToast = (data: ToastNotification) =>
+    client.setQueryData(TOAST_NOTIF_KEY, data);
+
+  useEffect(() => {
+    if (toast && timer === 0 && intervalId) {
+      window.clearInterval(intervalId);
+      setIntervalId(null);
+      window.clearTimeout(timeout.current);
+      timeout.current = window.setTimeout(
+        () => setToast(null),
+        IN_OUT_ANIMATION,
+      );
+    }
+  }, [timer, toast, intervalId]);
+
+  useEffect(
+    () => () => {
+      window.clearTimeout(timeout.current);
+    },
+    [],
+  );
+
+  useEffect(() => {
+    if (!toast) {
+      return;
+    }
+
+    window.clearInterval(intervalId);
+    setTimer(toast.timer);
+
+    if (!autoDismissNotifications && toast.onUndo && !intervalId) {
+      setIntervalId(TEMPORARY_ID);
+      return;
+    }
+
+    if (toast.timer <= 0) {
+      return;
+    }
+
+    setIntervalId(
+      window.setInterval(
+        () =>
+          setTimer((current) =>
+            INTERVAL_COUNT >= current ? 0 : current - INTERVAL_COUNT,
+          ),
+        INTERVAL_COUNT,
+      ),
+    );
+  }, [toast]);
+
+  const dismissToast = useCallback(() => {
+    if (!toast) {
+      return;
+    }
+
+    window.clearInterval(intervalId);
+    setToast({ ...toast, timer: 0 });
+  }, [toast, intervalId]);
+
+  const undoAction = async () => {
+    if (!toast?.onUndo) {
+      return;
+    }
+
+    await toast.onUndo();
+    setToast({ ...toast, timer: 0 });
+  };
+
+  useEffect(() => {
+    if (!isTouchDevice()) {
+      return;
+    }
+
+    const handler = () => dismissToast();
+    window.addEventListener('scroll', handler);
+    router.events.on('routeChangeStart', handler);
+
+    // eslint-disable-next-line consistent-return
+    return () => {
+      window.removeEventListener('scroll', handler);
+      router.events.off('routeChangeStart', handler);
+    };
+  }, [dismissToast]);
```

Comment:
This logic seems a bit too to maintain in one component.
We should consider refactoring to functions or hooks. It's hard to understand what these useEffects are all about without carefully following the code.

Reply:
The initialization of the toast notification is actually dependent to be a `useEffect` since we will have multiple instances of `useToastNotification`.

There are two options for us to work with the Notifications.
One is the current implementation.
The second is by rendering a toast component on each and every component where it will need the notification instead of the `useToastNotification` hook - from there only we can turn these ones into a function.

Perhaps I should add a comment on what the hooks actually do. What do you think?

Reply:
I love the high level implementation that you followed which is to have a single component and use a hook to trigger it.
My concern was very tactical, the fact that we have a lot of business logic inside the components makes it hard to read through. I suggest some refactoring around it. Just to make sure a new developer comes in and is able to understand what's going on